api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -171,31 +178,52 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
     } else if (workspace.getName().length() > 80) {
       throw new BadRequestException("Workspace name must be 80 characters or less");
     }
+
+    final boolean useBillingProjectBuffer = workbenchConfig.get().featureFlags.useBillingProjectBuffer;
+    if (useBillingProjectBuffer) {
+      if (Strings.isNullOrEmpty(workspace.getNamespace())) {
+        throw new BadRequestException("missing required field 'namespace'");
+      }
+      org.pmiops.workbench.db.model.Workspace existingWorkspace = workspaceService.getByName(
+              workspace.getNamespace(), workspace.getName());
+      if (existingWorkspace != null) {
+        throw new ConflictException(String.format(
+                "Workspace %s/%s already exists",
+                workspace.getNamespace(), workspace.getName()));
+      }
+    }
+
     User user = userProvider.get();
-    org.pmiops.workbench.db.model.Workspace existingWorkspace = workspaceService.getByName(
-        workspace.getNamespace(), workspace.getName());
-    if (existingWorkspace != null) {
-      throw new ConflictException(String.format(
-          "Workspace %s/%s already exists",
-          workspace.getNamespace(), workspace.getName()));
+    String workspaceNamespace;
+    if (useBillingProjectBuffer) {
+      BillingProjectBufferEntry bufferedBillingProject = billingProjectBufferService.assignBillingProject(user);
+      workspaceNamespace = bufferedBillingProject.getFireCloudProjectName();
+    } else {
+      workspaceNamespace = workspace.getNamespace();
     }
 
     // Note: please keep any initialization logic here in sync with CloneWorkspace().
-    FirecloudWorkspaceId workspaceId = generateFirecloudWorkspaceId(workspace.getNamespace(),
+    FirecloudWorkspaceId workspaceId = generateFirecloudWorkspaceId(workspaceNamespace,
         workspace.getName());
     FirecloudWorkspaceId fcWorkspaceId = workspaceId;
-    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace = null;
-    for (int attemptValue = 0; attemptValue < MAX_FC_CREATION_ATTEMPT_VALUES; attemptValue++) {
-      try {
-        fcWorkspace = attemptFirecloudWorkspaceCreation(fcWorkspaceId);
-        break;
-      } catch (ConflictException e) {
-        if (attemptValue >= 5) {
-          throw e;
-        } else {
-          fcWorkspaceId =
-              new FirecloudWorkspaceId(workspaceId.getWorkspaceNamespace(),
-                  workspaceId.getWorkspaceName() + Integer.toString(attemptValue));
+    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace;
+
+    if (useBillingProjectBuffer) {
+      fcWorkspace = attemptFirecloudWorkspaceCreation(fcWorkspaceId);
```

Comment:
I missed this in the previous review. There's a question of whether we should consider a billing project to be "assigned" before or after we add the user to the project.

If we fail on this line, we've assigned the project but it's effectively an orphan which we'd ideally clean up later.

But I think since that update could fail as well - I'm convincing myself that the way you have it now is probably ideal. We could also have another infrequent cron which cleans up any billing projects in this state (assigned, but don't have any associated workspace).